client/linuxabi/linux_abi.go

```diff
@@ -106,87 +112,185 @@ type SnpReportReq struct {
+// ABI returns the same object since it doesn't need a separate representation across the interface.
+func (r *SnpReportReqABI) ABI() BinaryConversion { return r }
+
+// Pointer returns a pointer to the object itself.
+func (r *SnpReportReqABI) Pointer() unsafe.Pointer {
+	return unsafe.Pointer(r)
+}
+
+// Finish is a no-op.
+func (r *SnpReportReqABI) Finish(b BinaryConvertible) error { return nil }
+
+// ABI returns the same object since it doesn't need a separate representation across the interface.
+func (r *SnpReportRespABI) ABI() BinaryConversion { return r }
+
+// Pointer returns a pointer to the object itself.
+func (r *SnpReportRespABI) Pointer() unsafe.Pointer {
+	return unsafe.Pointer(r)
+}
+
+// Finish checks the status of the message and translates it to a Golang error.
+func (r *SnpReportRespABI) Finish(b BinaryConvertible) error {
+	if r.Status != 0 {
+		switch r.Status {
+		case 0x16:
+			return errors.New("get_report had invalid parameters")
+		default:
+			return fmt.Errorf("unknown status: 0x%x", r.Status)
+		}
+	}
+	return nil
+}
+
+// SnpExtendedReportReqABI is Linux's sev-guest ioctl abi for sending a GET_EXTENDED_REPORT request.
+type SnpExtendedReportReqABI struct {
+	Data SnpReportReqABI
 
 	// Where to copy the certificate blob.
-	Certs []byte
+	CertsAddress unsafe.Pointer
 
 	// length of the certificate blob
 	CertsLength uint32
 }
 
-// SnpExtendedReportReq is Linux's sev-guest ioctl abi for sending a GET_EXTENDED_REPORT request.
+// SnpExtendedReportReq is close to Linux's sev-guest ioctl abi for sending a GET_EXTENDED_REPORT request,
+// but uses safer types for the Ioctl interface.
 type SnpExtendedReportReq struct {
-	Data SnpReportReq
+	Data SnpReportReqABI
 
-	// Where to copy the certificate blob.
-	CertsAddress uint64
+	// Certs receives the certificate blob after the extended report request.
+	Certs []byte
 
-	// length of the certificate blob
+	// CertsLength is the length of the certificate blob.
 	CertsLength uint32
 }
 
-// SnpUserGuestRequest is Linux's sev-guest ioctl abi for issuing a guest message.
-type SnpUserGuestRequest struct {
+// Pointer returns a pointer so the object itself.
+func (r *SnpExtendedReportReqABI) Pointer() unsafe.Pointer {
+	return unsafe.Pointer(r)
+}
+
+// Finish writes back the changed CertsLength value.
+func (r *SnpExtendedReportReqABI) Finish(b BinaryConvertible) error {
+	s, ok := b.(*SnpExtendedReportReq)
+	if !ok {
+		return fmt.Errorf("Finish argument is %v. Expects a *SnpExtendedReportReq", reflect.TypeOf(b))
+	}
+	s.CertsLength = r.CertsLength
+	return nil
+}
+
+// ABI returns an object that can cross the ABI boundary and copy back changes to the original
+// object.
+func (r *SnpExtendedReportReq) ABI() BinaryConversion {
+	var certsAddress unsafe.Pointer
+	if len(r.Certs) != 0 {
+		certsAddress = unsafe.Pointer(&r.Certs[0])
+	}
+	return &SnpExtendedReportReqABI{
+		Data:         r.Data,
+		CertsAddress: certsAddress,
+		CertsLength:  r.CertsLength,
```

Comment:
Should we check this? I guess it should be the same as len(r.Certs) -- right?

Reply:
No. the length may be larger if you're querying the kernel for the expected certs size.